Upgrade paper application to react form #162457306

akegan opened the pull request:
https://www.pivotaltracker.com/story/show/162457306

chrisdolendo committed:
[#162457306] - Init implement foundational react-final-form upgrade

chrisdolendo committed:
[#162457306] - Got record level error message working
- Need to organize files under final_form and finish error class assigning

akegan committed:
[#162457306] Get phone type working

chrisdolendo committed:
[#162457306] - Organize new react form files

chrisdolendo committed:
[#162457306] - Integrate AddressForm into react-final-form

chrisdolendo committed:
[#162457306] - Integrate react-final-form for alternate contact section

akegan committed:
[#162457306] Add second phone type to primary applicant

akegan committed:
[#162457306] Initial prototype of household members

akegan committed:
[#162457306] Add all household member fields

software-project committed:
[#162457306] Improve DOB validation

software-project committed:
[#162457306] Move DOB validation to validations file

software-project committed:
[#162457306] Add CheckboxField for eligibility section

chrisdolendo committed:
[#162457306] - Integrate react final form on household income and demographics section

chrisdolendo committed:
Merge branch 'chores/upgrade-react-form-#162457306' of https://github.com/Exygy/sf-dahlia-lap into chores/upgrade-react-form-#162457306

chrisdolendo committed:
[#162457306] - Integrate final form with Reserved Priority Section

software-project committed:
[#162457306] Add eligibility section for final form

software-project committed:
Merge branch 'chores/upgrade-react-form-#162457306' of https://github.com/Exygy/sf-dahlia-lap into chores/upgrade-react-form-#162457306

software-project committed:
[#162457306] Modify preferences section for final form

chrisdolendo committed:
[#162457306] - WIP Household Member section

software-project committed:
Merge branch 'chores/upgrade-react-form-#162457306' of https://github.com/Exygy/sf-dahlia-lap into chores/upgrade-react-form-#162457306

software-project committed:
[#162457306] WIP better DOB validation and fix for AddressForm

chrisdolendo committed:
[#162457306] - Agree to terms

software-project committed:
[#162457306] Add validation for household members

software-project committed:
[#162457306] Change preferences to FieldArray and add validation

software-project committed:
[#162457306] Add alternate contact validation. Fix linting issues.

software-project committed:
[#162457306] Fix tests and linting

software-project committed:
[#162457306] Fix e2e ApplicationNewPage test

software-project committed:
[#162457306] Update test snapshots

software-project committed:
[#162457306] Application page fixes for DOB and alerts

akegan committed:
[#162457306] Update snapshots

Update id and labels to have field name rather than "undefined"

software-project committed:
[#162457306] Fix labels, spacing and validation of preferences

software-project committed:
Merge branch 'chores/upgrade-react-form-#162457306' of https://github.com/Exygy/sf-dahlia-lap into chores/upgrade-react-form-#162457306

software-project committed:
[#162457306] Update test snapshot

akegan commented:
Starting to do an initial QA:

(not sure if you've already fixed these)
1. When saving application with validation, alert box shows up, but page doesn't scroll to the alert box
![image](https://user-images.githubusercontent.com/11825994/57658120-8fd09f80-7592-11e9-8ac5-7c375e1352b6.png)
2. Alert box doesn't close when "x" is clicked on 

3. HH member shows up as required even if none of the fields are populated. This does not match behavior on full 
![image](https://user-images.githubusercontent.com/11825994/57658302-4df42900-7593-11e9-9138-f3659519eb5a.png)

akegan commented:
Looks like we're missing the red border around fields for all validation errors

before (full):
![image](https://user-images.githubusercontent.com/11825994/57658738-d0311d00-7594-11e9-91f6-1e9ce1a1b56b.png)

after (this branch):
![image](https://user-images.githubusercontent.com/11825994/57658728-c60f1e80-7594-11e9-997d-05477f64ae61.png)

software-project committed:
[#162457306] Match household member validation with full, fix alert box and scroll

akegan committed:
[#162457306] Add red highlighting to invalid fields

akegan committed:
Merge branch 'chores/upgrade-react-form-#162457306' of https://github.com/Exygy/sf-dahlia-lap into chores/upgrade-react-form-#162457306

software-project committed:
Merge branch 'master' into chores/upgrade-react-form-#162457306

software-project committed:
[#162457306] Better DOB validation. Style fixes for preferences and errors

software-project committed:
Merge branch 'master' into chores/upgrade-react-form-#162457306

software-project committed:
[#162457306] Fix household member validation and test snapshot

akegan commented:
OK last QA comment.

We lost the behavior of checking that all 3 dob fields are touched before showing error. Now if you type in a month you get a red error saying you have an invalid date even if you haven't touched day or year yet
![image](https://user-images.githubusercontent.com/11825994/57735196-a2ada780-7658-11e9-8980-a6b84c473185.png)

I will try to fix after my next meeting (should be out by 3:45)

software-project committed:
[#162457306] Improve DOB validation

akegan commented:
oh dear @software-project dob validation looks worse now. I'll take a stab at fixing it

akegan commented:
we also appear to have lost email validation

akegan committed:
[#162457306] Update DOB validation

Only show error if all 3 fields have been touched

akegan committed:
[#162457306] Add email validation back to primary applicant

akegan committed:
[#162457306] Add comment for dob validation update

callaghanc commented on app/javascript/components/applications/application_form/AddressForm.js:
```diff
@@ -1,20 +1,20 @@
   // to do: refactor this to use props
   if (memberType === 'primaryApplicant') {
-    return fieldMap[fieldKey]
-  } else if (memberType === 'householdMember') {
```
@akegan looks like we used to expect "householdMember" as a value for `memberType` here, but now we don't - what's the context around that change?

akegan replied:
I didn't make this change, but it seems like it was an update to again accomodate the field array syntax. Maybe we should actually put this first, so the logic is: use the name if you have it, and fall back to applicant or alternate_contact if the name isn't defined.

@software-project do you have any more context?

software-project replied:
This is due to different way that forms are working with final form. HouseholdMembers are using FormArray. But in fact I can simplify it further.

callaghanc replied:
@software-project Ok that's a lot simpler! Could we also differentiate the var names `fieldName`, `fieldMap`, and `fieldKey` a little bit more? Seems like the field in `fieldName` is like a level higher than the field in `fieldMap` and `fieldKey`. If that's the case, it could be helpful to indicate that in the var names. Does something like `subFieldMap` and `subFieldKey` make sense? Or `addressFieldMap` and `addressFieldKey`? Something to show that there's this parent field with nested address fields.

callaghanc commented on app/javascript/components/applications/application_form/AddressForm.js:
```diff
@@ -1,20 +1,20 @@
 import { maxLengthMap } from '~/utils/formUtils'
 
-const buildField = (memberType, nestedField, fieldMap, fieldKey) => {
+const buildField = (memberType, name, fieldMap, fieldKey) => {
```
@akegan could we update the variable `name` to have a slightly more descriptive name? i'm not too familiar with the paper app form so i'm not sure what `name` is here. same for use of `name` in `AddressForm` component below.

akegan replied:
looks like name is used in the case of FieldArrays to pass down what the name is for form item in the value in the array. 

For example. for a household member, we'd expect it to be e.g. "household_members[0]", then the e.g. first name field would have the value "household_members[0].first_name" as it's "name". Maybe fieldName would make more sense?

software-project committed:
[#162457306] Refactor AddressForm

callaghanc commented on app/javascript/components/applications/application_form/preferences/PreferencesSection.js:
```diff
@@ -33,41 +34,44 @@ class PreferencesSection extends React.Component {
+                  </div>)
+              })
+              }
+              <div className='row'>
```
@software-project does the row with the Add Preference button need to be included within the `FieldArray` component?

software-project replied:
@callaghanc we paired on this one with Chris. We had had some hard times to make it work and moving it here worked. Maybe there is some other way, to keep it outside, but we couldn't find it.

akegan commented on spec/javascript/components/applications/__snapshots__/ApplicationEditPage.test.js.snap:
```diff
@@ -1134,12 +1134,12 @@ Array [
                               id="household_members_0_first_name"
                               maxLength={40}
-                              name="household_members[0].first_name"
+                              name="undefined.first_name"
```
@software-project looks like a bug was introduced here

software-project committed:
[#162457306] Refactor AddressForm

akegan commented:
@software-project looks like application save is totally broken now .

Doesn't look like your fault though:
```
<< API Error >> Faraday::ClientError, APEX_ERROR: API_LeaseUp_ApplicationService.RecordLockException: List index out of bounds: 0 
```

software-project commented:
> @software-project looks like application save is totally broken now .
> 
> Doesn't look like your fault though:
> 
> ```
> << API Error >> Faraday::ClientError, APEX_ERROR: API_LeaseUp_ApplicationService.RecordLockException: List index out of bounds: 0 
> ```

@akegan when are you getting this exception. It looks more like Salesforce issue.

akegan commented:
@software-project yeah it was salesforce issue. Fixed now

akegan commented:
heads up @software-project e2e tests are failing on application new page save

software-project commented:
@akegan issue is with latest changes to validation. Household member is not validated at all. Will try to fix.

software-project committed:
[#162457306] Fix household member validation

software-project committed:
[#162457306] Update test snapshot

callaghanc commented on app/javascript/components/applications/application_form/HouseholdMemberForm.js:
```diff
@@ -1,72 +1,72 @@
 } = formOptions
 
-const HouseholdMemberForm = ({ i, formApi }) => {
+const HouseholdMemberForm = ({fieldName, index, form}) => {
```
@software-project it's not clear here what type of information `fieldName` is. can we update the variable name or add a comment to further explain?

software-project replied:
@callaghanc what name do you find more descriptive? `formNamePrefix`, `fieldNamePrefix`? It's basically nested field name prefix. It's required for final for 'FieldArray'

callaghanc replied:
@software-project what values do we expect for `fieldName`? is it going to be "householdMember", or is it going to be the natural key?

software-project replied:
It's household member object name in the array, eg. `household_members[2]`

callaghanc replied:
ah ok thanks, so will the value of `fieldName` always be `` `household_members[${index}]` ``, similar to how it was before? or is there another possible value for `fieldName` that wouldn't be equal to `` `household_members[${index}]` ``?

software-project replied:
yes

callaghanc replied:
ok cool, then what do you think about passing ``fieldName={`household_members[${index}].middle_name`}`` to the fields in this form and removing the `fieldName` prop from `HouseholdMemberForm` entirely? that makes one less thing to think about here, since `fieldName` isn't really bringing any info that isn't already in `index`

software-project replied:
that would work, sure.

callaghanc replied:
cool i'll push that change

callaghanc commented on app/javascript/utils/form/validations.js:
```diff
@@ -102,4 +102,26 @@ validate.any = (...fns) => (value) => {
       map(fns, fn => fn(value))))
 }
 
+validate.isValidDOB = (person, errors, isPrimaryApplicant = false) => {
```
@software-project how would you feel about using an options object for the `isPrimaryApplicant` argument? e.g. `opts = {}`. then the call is `isValidDOB(person, errors, {isPrimaryApplicant: true})`, which indicates what that last argument is for, instead of `isValidDOB(person, errors, true)`, which doesn't indicate what `true` will do and makes the dev have to go look up the function

callaghanc committed:
[162457306] WIP - refactor labelize function for clarity

Also standardize naming of index variables in the preference form, and remove an unused variable from a function in the preference form. This commit marked WIP because I was unable to test it due to Salesforce being down. Pushing for other team members to review in the meantime. Will test when Salesforce is back up.

software-project committed:
[#162457306] Refactor Fields and listing preference L_W

software-project committed:
[#162457306] Refactor Fields and listing preference L_W

software-project committed:
Merge branch 'chores/upgrade-react-form-#162457306' of https://github.com/Exygy/sf-dahlia-lap into chores/upgrade-react-form-#162457306

software-project committed:
[#162457306] Update test snapshots

akegan commented:
@software-project looks like e2e test for selecting live/work pref is broken. Also branch is out of date with master

software-project committed:
[#162457306] Fix index missing on PreferenceForm

software-project committed:
[#162457306] Fix index missing on PreferenceForm

software-project committed:
Merge branch 'master' into chores/upgrade-react-form-#162457306

akegan commented:
I can't officially approve because I opened the PR but just did another round of solid QA and it loooks good!!! let's merge! @callaghanc @software-project you can approve and merge!

callaghanc committed:
[#162457306] Streamline props of HouseholdMemberForm

akegan commented:
OK did some more QA around HH members and still looks good! @callaghanc please approve and merge! 🎉

callaghanc reviewed (approved):
approved based on @akegan's qa
